File: UsingFMDB-Objective-C/UsingFMDB-Objective-C/BookDAO.m
//
//  BookDAO.m
//  UsingFMDB-Objective-C
//
//  Created by akabeko on 2016/12/03.
//  Copyright © 2016年 akabeko. All rights reserved.
//

#import "BookDAO.h"
#import "Book.h"
#import <FMDatabase.h>
#import <FMResultSet.h>

/** Query for the create table. */
static NSString * const kSQLCreate = @""
"CREATE TABLE IF NOT EXISTS books ("
  "id INTEGER PRIMARY KEY AUTOINCREMENT, "
  "author TEXT, "
  "title TEXT, "
  "release_date INTEGER"
");";

/** Query for the select rows. */
static NSString * const kSQLSelect = @""
"SELECT "
  "id, author, title, release_date "
"FROM "
  "books;"
"ORDER BY "
  "author, title;";

/** Query for the inssert row. */
static NSString * const kSQLInsert = @""
"INSERT INTO "
  "books (author, title, release_date) "
"VALUES "
  "(?, ?, ?);";

/** Query for the update row. */
static NSString * const kSQLUpdate = @""
"UPDATE "
  "books "
"SET "
  "author = ?, title = ?, release_date = ? "
"WHERE "
  "id = ?;";

/** Query for the delete row. */
static NSString * const kSQLDelete = @""
"DELETE FROM books WHERE id = ?;";

@interface BookDAO ()

/** Instance of the database connection. */
@property (nonatomic) FMDatabase *db;

@end

@implementation BookDAO

/**
 * Initialize the instance.
 *
 * @param db Instance of the database connection.
 *
 * @return Instance.
 */
- (instancetype)init:(FMDatabase *)db {
    self = [super init];
    if (self && db) {
        self.db = db;
    }

    return self;
}

/**
 * Deallocates the memory occupied by the receiver.
 */
- (void)dealloc {
    [self.db close];
}

/**
 * Create the table.
 *
 *
 * @return YES if successful.
 */
- (BOOL)create {
    return [self.db executeUpdate:kSQLCreate];
}

/**
 * Add a book.
 *
 * @param author      Author.
 * @param title       Title.
 * @param releaseDate Release date.
 *
 * @return Added book.
 */
- (Book *)add:(NSString *)author title:(NSString *)title releaseDate:(NSDate *)releaseDate {
    Book *book = nil;
    if ([self.db executeUpdate:kSQLInsert, author, title, releaseDate]) {
        NSInteger bookId = [self.db lastInsertRowId];
        book = [Book bookWithId:bookId author:author title:title releaseDate:releaseDate];
    }

    return book;
}

/**
 * Read a books.
 *
 * @return Readed books.
 */
- (NSArray *)read {
    NSMutableArray *books = [NSMutableArray arrayWithCapacity:0];
    FMResultSet    *results = [self.db executeQuery:kSQLSelect];

    while ([results next]) {
        [books addObject:[Book bookWithId:[results intForColumnIndex:0]
                                   author:[results stringForColumnIndex:1]
                                    title:[results stringForColumnIndex:2]
                              releaseDate:[results dateForColumnIndex:3]]];
    }

    return books;
}

/**
 * Remove a book.
 *
 * @param bookId The identifier of the book to remove.
 *
 * @return YES if successful.
 */
- (BOOL)remove:(NSInteger)bookId {
    return [self.db executeUpdate:kSQLDelete, [NSNumber numberWithInteger:bookId]];
}

/**
 * Update a book.
 *
 * @param book The data of the book to update.
 *
 * @return YES if successful.
 */
- (BOOL)update:(Book *)book {
    return [self.db executeUpdate:kSQLUpdate,
            book.author,
            book.title,
            book.releaseDate,
            [NSNumber numberWithInteger:book.bookId]];
}

@end

File: UsingFMDB-Swift/UsingFMDB-Swift/BookDAO.swift
//
//  BookDAO.swift
//  UsingFMDB-Swift
//
//  Created by akabeko on 2016/12/17.
//  Copyright © 2016年 akabeko. All rights reserved.
//

import UIKit
import FMDB

/// Manage for the books data table.
class BookDAO: NSObject {
    /// Query for the create table.
    private static let SQLCreate = "" +
    "CREATE TABLE IF NOT EXISTS books (" +
      "id INTEGER PRIMARY KEY AUTOINCREMENT, " +
      "author TEXT, " +
      "title TEXT, " +
      "release_date INTEGER" +
    ");"

    /// Query for the inssert row.
    private static let SQLSelect = "" +
    "SELECT " +
      "id, author, title, release_date " +
    "FROM " +
      "books;" +
    "ORDER BY " +
      "author, title;"

    /// Query for the inssert row.
    private static let SQLInsert = "" +
    "INSERT INTO " +
      "books (author, title, release_date) " +
    "VALUES " +
      "(?, ?, ?);"

    /// Query for the update row.
    private static let SQLUpdate = "" +
    "UPDATE " +
      "books " +
    "SET " +
      "author = ?, title = ?, release_date = ? " +
    "WHERE " +
      "id = ?;"

    /// Query for the delete row.
    private static let SQLDelete = "DELETE FROM books WHERE id = ?;"

    /// Instance of the database connection.
    private let db: FMDatabase

    /// Initialize the instance.
    ///
    /// - Parameter db: Instance of the database connection.
    init(db: FMDatabase) {
        self.db = db
        super.init()
    }

    deinit {
        self.db.close()
    }

    /// Create the table.
    func create() {
        self.db.executeUpdate(BookDAO.SQLCreate, withArgumentsIn: nil)
    }

    /// Add the book.
    ///
    /// - Parameters:
    ///   - author:      Author.
    ///   - title:       Title.
    ///   - releaseDate: Release date.
    /// - Returns: Added the book.
    func add(author: String, title: String, releaseDate: Date) -> Book? {
        var book: Book? = nil
        if self.db.executeUpdate(BookDAO.SQLInsert, withArgumentsIn: [author, title, releaseDate]) {
            let bookId = db.lastInsertRowId()
            book = Book(bookId: Int(bookId), author: author, title: title, releaseDate: releaseDate)
        }

        return book
    }

    /// Read a books.
    ///
    /// - Returns: Readed books.
    func read() -> Array<Book> {
        var books = Array<Book>()
        if let results = self.db.executeQuery(BookDAO.SQLSelect, withArgumentsIn: nil) {
            while results.next() {
                let book = Book(bookId: results.long(forColumnIndex: 0),
                                author: results.string(forColumnIndex: 1),
                                title: results.string(forColumnIndex: 2),
                                releaseDate: results.date(forColumnIndex: 3))
                books.append(book)
            }
        }

        return books
    }

    /// Remove a book.
    ///
    /// - Parameter bookId: The identifier of the book to remove.
    /// - Returns: "true" if successful.
    func remove(bookId: Int) -> Bool {
        return self.db.executeUpdate(BookDAO.SQLDelete, withArgumentsIn: [bookId])
    }

    /// Update a book.
    ///
    /// - Parameter book: The data of the book to update.
    /// - Returns: "true" if successful.
    func update(book: Book) -> Bool {
        return db.executeUpdate(BookDAO.SQLUpdate,
                                withArgumentsIn: [
                                    book.author,
                                    book.title,
                                    book.releaseDate,
                                    book.bookId])
    }
}
